File: CommonCode/SpectrumGenerator.m
//
//  SpectrumGenerator.m
//  Spectrogeddon
//
//  Created by Tom York on 14/04/2014.
//  
//

#import "SpectrumGenerator.h"
#import "FastFourierTransform.h"
#import "AudioSource.h"
#import "TimeSequence.h"
#import "DisplaySettings.h"

@interface SpectrumGenerator ()
@property (nonatomic,strong) FastFourierTransform* transformer;
@property (nonatomic,strong) AudioSource* audioSource;
@property (nonatomic,strong) dispatch_queue_t transformQueue;
@end

@implementation SpectrumGenerator

+ (NSDictionary*)availableSources
{
    return [AudioSource availableAudioSources];
}

- (instancetype)init
{
    if((self = [super init]))
    {
        _transformer = [[FastFourierTransform alloc] init];
        _transformQueue = dispatch_queue_create("com.spectrogeddon.fft", DISPATCH_QUEUE_SERIAL);
        
        typeof(self) __weak weakSelf = self;
        _audioSource = [[AudioSource alloc] initWithNotificationQueue:_transformQueue block:^(NSArray* channels) {
            SpectrumGenerator* strongSelf = weakSelf;
            if(!strongSelf)
            {
                return;
            }

            NSMutableArray* spectrums = [[NSMutableArray alloc] init];
            for(TimeSequence* oneTimeSequence in channels)
            {
                TimeSequence* fft = [strongSelf.transformer transformSequence:oneTimeSequence];
                if(!fft)
                {
                    return;
                }
                [spectrums addObject:fft];
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                
                [strongSelf.delegate spectrumGenerator:self didGenerateSpectrums:spectrums];
            });
            
        }];
    }
    return self;
}

- (void)startGenerating
{
    [self.audioSource startCapturing];
}

- (void)stopGenerating
{
    [self.audioSource stopCapturing];
}

- (void)useSettings:(DisplaySettings*)settings
{
    NSString* const audioSourceId = settings.preferredAudioSourceId;
    if(audioSourceId != nil && [[[self class] availableSources].allValues containsObject:audioSourceId]) {
        self.audioSource.preferredAudioSourceID = audioSourceId;
    }
    self.audioSource.bufferSizeDivider = settings.sharpness;
}

@end

File: CommonCode/SpectrumGenerator.swift
//
//  SpectrumGenerator.swift
//  SpectrogeddonOSX
//
//  Created by Tom York on 14/05/2019.
//  Copyright © 2019 Spectrogeddon. All rights reserved.
//
import Foundation

public protocol SpectrumGeneratorDelegate: AnyObject {
    // Marking just this function as @MainActor ensures that whoever receives
    // the audio data (UI/Renderers) processes it safely on the main thread.
    @MainActor func spectrumGenerator(_ generator: SpectrumGenerator, didGenerate spectrums: [TimeSequence])
}

// Reverted @MainActor and added @unchecked Sendable
public class SpectrumGenerator: @unchecked Sendable {
    
    public typealias SourceName = String
    public typealias SourceID = String
    
    public static var availableSources: [SourceName: SourceID] {
        return AudioSource.availableAudioSources
    }

    public weak var delegate: SpectrumGeneratorDelegate?
    
    private let transformer: FastFourierTransform
    private let queue: DispatchQueue
    private var audioSource: AudioSource!

    public init() throws {
        transformer = FastFourierTransform()
        queue = DispatchQueue(label: "fft", qos: .default)
        
        audioSource = try AudioSource(notificationQueue: queue) { [weak self] channels in
            guard let strongSelf = self else { return }
            
            // 1. Heavy FFT math processing runs sequentially on your background queue
            let spectrums: [TimeSequence] = channels.compactMap {
                return strongSelf.transformer.transform($0)
            }
            
            // 2. Dispatch data back to the main thread cleanly, bypassing
            // the brittle runtime Actor thunks that caused the crash.
            DispatchQueue.main.async {
                strongSelf.delegate?.spectrumGenerator(strongSelf, didGenerate: spectrums)
            }
        }
    }
    
    public func start() {
        audioSource.startCapturing()
    }
    
    public func stop() {
        audioSource.stopCapturing()
    }
    
    public func useSettings(_ settings: DisplaySettings) {
        if let sourceId = settings.preferredAudioSourceId, SpectrumGenerator.availableSources.values.contains(sourceId) {
            audioSource.preferredAudioSourceId = sourceId
        }
        audioSource.bufferSizeDivider = settings.sharpness
    }
}
